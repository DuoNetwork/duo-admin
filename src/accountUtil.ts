/*
This program is to create parity accounts.
Fuel created accounts with Ether or collect ether from created accoutns
Use created accounts to interact with smart contract
*/
import * as fs from 'fs';
import * as CST from './constants';
import ContractUtil from './contractUtil';
import { IAccount, IOption } from './types';
import util from './util';
const Tx = require('ethereumjs-tx');
const accountsFile = './src/static/KovanAccounts.json';
const schedule = require('node-schedule');

export class AccountUtil {
	public async transferEth(
		contractUtil: ContractUtil,
		from: string,
		privatekey: string,
		to: string,
		amt: number,
		nonce: number
	) {
		// util.log(
		// 	'transfering from: ' + from + ' to: ' + to + ' amount: ' + amt + ' nonce ' + nonce
		// );
		const privateKey = new Buffer(privatekey.toLocaleLowerCase(), 'hex');
		const rawTx = {
			nonce: nonce,
			gasPrice: contractUtil.web3.utils.toHex(
				(await contractUtil.getGasPrice()) || CST.DEFAULT_GAS_PRICE
			),
			gasLimit: contractUtil.web3.utils.toHex(23000),
			from: from,
			to: to,
			value: contractUtil.web3.utils.toHex(
				contractUtil.web3.utils.toWei(util.truncateNum(amt, 3) + '', 'ether')
			)
		};
		const tx = new Tx(rawTx);
		tx.sign(privateKey);
		const serializedTx = tx.serialize();
		await contractUtil.web3.eth
			.sendSignedTransaction('0x' + serializedTx.toString('hex'))
			.on('receipt', console.log);
	}

	public getMainAccount(option: IOption): IAccount {
		let mainAccount;
		let privateKey;
		if (!option.from || !option.privateKey) {
			mainAccount = CST.KOVAN_ACCOUNTS[0].address;
			privateKey = CST.KOVAN_ACCOUNTS[0].privateKey;
			console.log('no from account specified, use defaut account: ' + mainAccount);
		} else {
			mainAccount = option.from;
			privateKey = option.privateKey;
		}
		return {
			address: mainAccount,
			privateKey: privateKey
		};
	}
	public async createAccount(contractUtil: ContractUtil, option: IOption) {
		// by default, only one account is created
		const saveAccount = option.saveAccount ? 'yes' : 'no';
		console.log('creating accounts: ' + option.accountNum + ' saveAccounts: ' + saveAccount);
		const data = fs.readFileSync(accountsFile, 'utf8');
		const accountsData = JSON.parse(data);
		// console.log(accountsData);
		let createdAccount;
		for (let i = 0; i < option.accountNum; i++) {
			console.log('creating account no: ' + i);
			createdAccount = await contractUtil.web3.eth.accounts.create(
				contractUtil.web3.utils.randomHex(32)
			);
			console.log(createdAccount);
			if (option.saveAccount) {
				const obj = {
					address: createdAccount.address,
					privateKey: createdAccount.privateKey
				};
				accountsData.push(obj);
			}
		}
		// console.log(accountsData);
		const allAccounts = JSON.stringify(accountsData);
		fs.writeFileSync(accountsFile, allAccounts, 'utf8');
	}

	public async allAccountsInfo(contractUtil: ContractUtil) {
		const data = fs.readFileSync(accountsFile, 'utf8');
		const accountsData = JSON.parse(data);
		console.log('there are total accounts no ' + accountsData.length);
		accountsData.forEach(async account => {
			const balance = await contractUtil.web3.eth.getBalance(account.address);
			console.log(
				'account address: ' +
					account.address +
					' eth balance ' +
					contractUtil.web3.utils.fromWei(balance + '', 'ether')
			);
		});
	}

	public async fuelAccounts(contractUtil: ContractUtil, option: IOption) {
		const mainAccount: IAccount = this.getMainAccount(option);

		let mainAccountBalance = await contractUtil.web3.eth.getBalance(mainAccount.address);
		mainAccountBalance = Number(
			contractUtil.web3.utils.fromWei(mainAccountBalance + '', 'ether')
		);
		if (mainAccountBalance < option.total) {
			console.log(
				'mian account balance is: ' + mainAccountBalance + ' less than 10 ether, stop'
			);
		} else {
			console.log(
				'mian account balance is: ' + mainAccountBalance + ' starting fuel other accounts'
			);
			const data = fs.readFileSync(accountsFile, 'utf8');
			const accountsData = JSON.parse(data);
			const avgEthPerAccount = option.total / accountsData.length;

			const filterredAccounts: any[] = [];
			let promiseList: any[] = [];
			promiseList = accountsData.map(async account => {
				let currentBalance = Number(
					await contractUtil.web3.eth.getBalance(account.address)
				);
				currentBalance = Number(
					contractUtil.web3.utils.fromWei(currentBalance + '', 'ether')
				);
				// console.log(currentBalance, option.minEther);

				if (currentBalance < option.minEther) {
					return filterredAccounts.push(account);
					// promiseList.push()
				} else {
					return undefined;
				}
			});
			await Promise.all(promiseList);

			if (filterredAccounts.length > 0) {
				console.log('need fuel ether to ' + filterredAccounts.length + ' address');
				const sendInterval = 10;
				const startTime = new Date(Date.now());
				const endTime = new Date(
					startTime.getTime() +
						sendInterval * 1000 * (filterredAccounts.length + 1) +
						1000
				);
				let i = 0;
				let nonce: number = await contractUtil.web3.eth.getTransactionCount(
					mainAccount.address
				);
				const rule = '*/' + sendInterval + ' * * * * *';
				schedule.scheduleJob({ start: startTime, end: endTime, rule: rule }, async () => {
					if (i >= filterredAccounts.length) {
						console.log('completd fuel process');
					} else {
						i++;
						const account = filterredAccounts[i - 1];
						const randomAmt = Math.random() * avgEthPerAccount;
						const amt = randomAmt < 0.01 ? 0.01 : randomAmt;
						util.log(
							'starting transfer ether to account no ' +
								i +
								' address: ' +
								account.address +
								' ethAmt: ' +
								amt
						);
						nonce++;
						await this.transferEth(
							contractUtil,
							mainAccount.address,
							mainAccount.privateKey,
							account.address,
							amt,
							nonce - 1
						);
					}
				});
			} else {
				console.log('no need to fuel');
			}
		}
	}

	public async collectEther(contractUtil: ContractUtil, option: IOption) {
		const mainAccount: IAccount = this.getMainAccount(option);
		console.log(mainAccount);
		const data = fs.readFileSync(accountsFile, 'utf8');
		const accountsData = JSON.parse(data);

		const filterredAccounts: any[] = [];
		let promiseList: any[] = [];
		promiseList = accountsData.map(async account => {
			const currentBalance = Number(await contractUtil.web3.eth.getBalance(account.address));
			if (currentBalance > CST.TRANSFER_GAS_TH) {
				return filterredAccounts.push(account);
				// promiseList.push()
			} else {
				return undefined;
			}
		});
		await Promise.all(promiseList);
		if (filterredAccounts.length > 0) {
			util.log('need collect ether form ' + filterredAccounts.length + ' accounts');
			const sendInterval = 10;
			const startTime = new Date(Date.now());
			const endTime = new Date(
				startTime.getTime() + sendInterval * 1000 * (filterredAccounts.length + 1) + 1000
			);
			let i = 0;
			const rule = '*/' + sendInterval + ' * * * * *';
			schedule.scheduleJob({ start: startTime, end: endTime, rule: rule }, async () => {
				if (i >= filterredAccounts.length) {
					console.log('completd collect process');
				} else {
					i++;
					const account = filterredAccounts[i - 1];
					const nonce = await contractUtil.web3.eth.getTransactionCount(account.address);
					const currentBalance = Number(
						await contractUtil.web3.eth.getBalance(account.address)
					);
					const amt: number = Number(
						contractUtil.web3.utils.fromWei(
							currentBalance - CST.TRANSFER_GAS_TH + '',
							'ether'
						)
					);
					util.log(
						'starting collect ether from account no ' +
							i +
							' address: ' +
							account.address +
							' ethAmt: ' +
							amt
					);
					await this.transferEth(
						contractUtil,
						account.address,
						account.privateKey.replace('0x', ''),
						mainAccount.address,
						amt,
						nonce
					);
				}
			});
		} else {
			util.log('no account to collect ether from');
		}
	}

	public async makeCreation(contractUtil: ContractUtil, option: IOption) {
		const data = fs.readFileSync(accountsFile, 'utf8');
		const accountsData = JSON.parse(data);
		const alpha = option.alpha;
		const accountsIdxToCreate = util.generateRandomIdx(accountsData.length - 1, alpha);

		const filterredAccounts: any[] = [];
		let promiseList: any[] = [];
		promiseList = accountsIdxToCreate.map(async idx => {
			const currentBalance = Number(
				await contractUtil.web3.eth.getBalance(accountsData[idx].address)
			);
			if (currentBalance > CST.TRANSFER_GAS_TH) {
				return filterredAccounts.push(accountsData[idx]);
			} else {
				return undefined;
			}
		});
		await Promise.all(promiseList);
		console.log(filterredAccounts);

		util.log('there are ' + filterredAccounts.length + ' accounts to create');
		const sendInterval = 10;
		const startTime = new Date(Date.now());
		const endTime = new Date(
			startTime.getTime() + sendInterval * 1000 * (filterredAccounts.length + 1) + 1000
		);
		let i = 0;
		const rule = '*/' + sendInterval + ' * * * * *';
		schedule.scheduleJob({ start: startTime, end: endTime, rule: rule }, async () => {
			if (i >= filterredAccounts.length) {
				console.log('completd creation process');
			} else {
				i++;
				const account = accountsData[i - 1];
				const nonce = await contractUtil.web3.eth.getTransactionCount(account.address);
				const currentBalance = Number(
					await contractUtil.web3.eth.getBalance(account.address)
				);
				console.log(currentBalance - CST.TRANSFER_GAS_TH);
				const amt: number =
					Number(
						contractUtil.web3.utils.fromWei(
							currentBalance - CST.TRANSFER_GAS_TH + '',
							'ether'
						)
					) * Math.random();
				util.log(
					'starting creating from account no ' +
						i +
						' address: ' +
						account.address +
						' ethAmt: ' +
						amt
				);
				option.address = account.address;
				option.privateKey = account.privateKey.replace('0x', '');
				option.eth = util.truncateNum(amt);
				option.gasLimit = 400000;
				await contractUtil.create(option, nonce);
			}
		});
	}

	public async makeRedemption(contractUtil: ContractUtil, option: IOption) {
		//
		const data = fs.readFileSync(accountsFile, 'utf8');
		const accountsData = JSON.parse(data);
		const filterredAccounts: any[] = [];
		let promiseList: any[] = [];
		promiseList = accountsData.map(async account => {
			const balanceOfA = Number(
				await contractUtil.contract.methods.balanceOf(0, account.address).call()
			);
			const balanceOfB = Number(
				await contractUtil.contract.methods.balanceOf(1, account.address).call()
			);
			const currentBalance = Number(
				await contractUtil.web3.eth.getBalance(account.address)
			);
			// console.log(balanceOfA, balanceOfB);
			if (balanceOfA > 0 && balanceOfB > 0 && currentBalance > CST.REDEEM_GAS_TH) {
				return filterredAccounts.push(account);
				// promiseList.push()
			} else {
				return undefined;
			}
		});
		await Promise.all(promiseList);
		// console.log(filterredAccounts);
		// const alpha = option.alpha;
		// const accountsIdxToCreate = util.generateRandomIdx(accountsData.length - 1, alpha);

		util.log('there are ' + filterredAccounts.length + ' accounts to Redeem');
		const sendInterval = 10;
		const startTime = new Date(Date.now());
		const endTime = new Date(
			startTime.getTime() + sendInterval * 1000 * (filterredAccounts.length + 1) + 1000
		);
		let i = 0;
		const rule = '*/' + sendInterval + ' * * * * *';
		schedule.scheduleJob({ start: startTime, end: endTime, rule: rule }, async () => {
			if (i >= filterredAccounts.length) {
				console.log('completd redemption process');
			} else {
				i++;
				const account = filterredAccounts[i - 1];
				const nonce = await contractUtil.web3.eth.getTransactionCount(account.address);
				const balanceOfA = Number(
					(
						(await contractUtil.contract.methods.balanceOf(0, account.address).call()) *
						Math.random()
					).toFixed(0)
				);
				const balanceOfB = Number(
					(
						(await contractUtil.contract.methods.balanceOf(1, account.address).call()) *
						Math.random()
					).toFixed(0)
				);
				util.log(
					'starting redemption from account no ' +
						i +
						' address: ' +
						account.address +
						' amtA: ' +
						balanceOfA +
						' amtB: ' +
						balanceOfB
				);
				option.address = account.address;
				option.privateKey = account.privateKey.replace('0x', '');
				option.amtA = balanceOfA;
				option.amtB = balanceOfB;
				option.gasLimit = 800000;
				console.log(option);
				await contractUtil.redeem(option, nonce);
			}
		});
	}
}

const accountUtil = new AccountUtil();
export default accountUtil;
