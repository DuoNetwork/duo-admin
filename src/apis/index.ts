import * as CST from '../constants';
import BaseApi from './BaseApi';
import bitfinexApi from './bitfinexApi';
import gdaxApi from './gdaxApi';
import geminiApi from './geminiApi';
import krakenApi from './krakenApi';

const apis: { [key: string]: BaseApi } = {
	[CST.EXCHANGE_BITFINEX]: bitfinexApi,
	[CST.EXCHANGE_GEMINI]: geminiApi,
	[CST.EXCHANGE_KRAKEN]: krakenApi,
	[CST.EXCHANGE_GDAX]: gdaxApi
};

export default apis;
