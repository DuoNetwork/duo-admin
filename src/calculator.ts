import sqlUtil from './sqlUtil';
import * as CST from './constants';
import { Price, Trade } from './types';

export class Calculateor {
	getTradesForInterval(trades: Trade[], currentTimestamp: number, interval: number): Trade[] {
		const oneInterval = 5 * 60 * 1000;
		const upperTime: number = currentTimestamp - interval * oneInterval;
		const lowerTime: number = upperTime - oneInterval;

		return trades.filter(
			trade =>
				Number(trade.timestamp) <= upperTime &&
				Number(trade.timestamp) > lowerTime
		);
	}

	getVolumeMedianPrice(trades: Trade[], timestamp: number): Price {
		trades.sort((a, b) => Number(a.price) - Number(b.price));
		const cumVols: number[] = [];
		let cumVol = 0;
		trades.forEach((trade, index) => {
			cumVol += Number(trade.amount);
			cumVols[index] = cumVol;
		});
		const halfTotalVol: number = cumVol / 2;
		const medianIndex = cumVols.findIndex(v => v >= halfTotalVol);
		return { price: Number(trades[medianIndex].price), volume: cumVol, timestamp: timestamp };
	}

	getExchangePriceFix(trades: Trade[], currentTimestamp: number): Price {
		for (let i = 0; i < 12; i++) {
			const subTrades = this.getTradesForInterval(trades, currentTimestamp, i);
			if (subTrades.length > 0) return this.getVolumeMedianPrice(subTrades, currentTimestamp);
		}

		return { price: 0, volume: 0, timestamp: currentTimestamp };
	}

	getWeights(rawVolume: number[]): number[] {
		const totalVol = rawVolume.reduce((a, b) => a + b);
		const weightArray = rawVolume.map(v => v / totalVol);
		return weightArray;
	}

	validateWeights(weights: number[]): boolean {
		const numOfValidExchanges = weights.length;
		for (let i = 0; i < weights.length; i++)
			if (weights[i] > CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges][i]) return false;
		return true;
	}

	modifyWeights(weights: number[]): number[] {
		const numOfValidExchanges = weights.length;
		let isValid: boolean = this.validateWeights(weights);
		while (!isValid) {
			let sumOfCapped: number = 0;
			let sumOfUnCapped: number = 0;
			const weightCaps = CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges];
			for (let i = 0; i < weights.length; i++) {
				if (weights[i] >= weightCaps[i]) {
					weights[i] = weightCaps[i];
					sumOfCapped += weightCaps[i];
				} else {
					sumOfUnCapped += weights[i];
				}
			}

			for (let i = 0; i < weights.length; i++) {
				if (weights[i] < weightCaps[i]) {
					weights[i] = weights[i] / sumOfUnCapped * (1 - sumOfCapped);
				}
			}
			// console.log(weights);
			isValid = this.validateWeights(weights);
		}
		return weights;
	}

	consolidatePriceFix(exchangePriceVolume: Array<Price>): number {
		const filterredExchanges = exchangePriceVolume.filter(item => item.volume > 0);

		// sort based on volume from large to small
		filterredExchanges.sort((a, b) => b.volume - a.volume);
		const volumeList = filterredExchanges.map(item => item.volume);

		if (filterredExchanges.length === 0)
			// use previous priceFix
			return 0;
		else if (filterredExchanges.length === 1)
			// let finalArray: number[] = [];
			return filterredExchanges[0].price;
		else {
			console.log('there are ' + filterredExchanges.length + ' valid exchanges');
			const totalVol = volumeList.reduce((a, b) => a + b);
			const weights = volumeList.map(v => v / totalVol);
			const finalWeights: number[] = this.modifyWeights(weights);
			return finalWeights.reduce(
				(accum, w, index) => accum + exchangePriceVolume[index].price * w
			);
		}
	}

	async getPriceFix(): Promise<Price> {
		const currentTimestamp: number = Math.floor(Date.now());
		const trades = await sqlUtil.readSourceData(currentTimestamp);
		const EXCHANGES_TRADES: { [key: string]: Trade[] } = {
			[CST.EXCHANGE_BITFINEX]: [],
			[CST.EXCHANGE_GEMINI]: [],
			[CST.EXCHANGE_GDAX]: [],
			[CST.EXCHANGE_KRAKEN]: []
		};

		trades.forEach(item => {
			if (item.source === CST.EXCHANGE_BITFINEX)
				EXCHANGES_TRADES[CST.EXCHANGE_BITFINEX].push(item);
			else if (item.source === CST.EXCHANGE_GEMINI)
				EXCHANGES_TRADES[CST.EXCHANGE_GEMINI].push(item);
			else if (item.source === CST.EXCHANGE_GDAX)
				EXCHANGES_TRADES[CST.EXCHANGE_GDAX].push(item);
			else if (item.source === CST.EXCHANGE_KRAKEN)
				EXCHANGES_TRADES[CST.EXCHANGE_KRAKEN].push(item);
		});

		const exchangePriceVolume = CST.EXCHANGES.map(src =>
			this.getExchangePriceFix(EXCHANGES_TRADES[src], currentTimestamp)
		);

		const priceFix: number = this.consolidatePriceFix(exchangePriceVolume);

		if (priceFix === 0) {
			console.log('no priceFix found, use the last ETH price');
			const lastPriceObj = await sqlUtil.readLastPrice();
			console.log(
				'the priceFix is: ' + lastPriceObj.price + ' at timestamp ' + lastPriceObj.timestamp
			);
			return lastPriceObj;
		} else {
			console.log('the priceFix is: ' + priceFix + ' at timestamp ' + currentTimestamp);
			console.log(exchangePriceVolume);
			console.log(exchangePriceVolume.reduce((sum, p) => sum + p.volume, 0));
			const priceObj = {
				price: priceFix,
				volume: exchangePriceVolume.reduce((sum, p) => sum + p.volume, 0),
				timestamp: currentTimestamp
			};
			// save price into DB
			await sqlUtil.insertPrice(priceObj);

			return priceObj;
		}
	}
}
const calculator = new Calculateor();
export default calculator;
