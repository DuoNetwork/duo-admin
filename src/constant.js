export * from '../../sachiel/src/functions/common/constants';
export const REDRAW_UPDATE_THRESHOLD = 10;

//export const provider = 'https://mainnet.infura.io/Ky03pelFIxoZdAUsr82w';
export const provider = 'https://kovan.infura.io/WSDscoNUvMiL1M7TvMNP ';
//export const provider = 'http://localhost:8545';

export const EXCHANGE_BITFINEX = 'BITFINEX';
export const EXCHANGE_GEMINI = 'GEMINI';
export const EXCHANGE_KRAKEN = 'KRAKEN';
export const EXCHANGE_GDAX = 'GDAX';

//db setting
export const DB_HOST = 'localhost';
export const DB_USER = 'root';
export const DB_PASSWORD = '';
export const DB_PRICEFEED = 'priceFeedDB';
export const DB_TABLE_TRADE = 'ETH_Trades_Table';

export const addressCustodianContract = '0x4458294e6d83696ec9a400dc4f6c27c9a0a1c559';
export const pfAddress = '0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a';
export const privateKey = '5e02a6a6b05fe971309cba0d0bd8f5e85f25e581d18f89eb0b6da753d18aa285';
export const ETH_PRICE_LINK = 'https://min-api.cryptocompare.com/data/price?fsym=ETH&tsyms=USD';

export const ETHSCAN_API_KEY = '8VS7KBP65Q7TQE4NGNUDEF69M6M6TH4BRS';
export const ETHSCAN_API_KOVAN_LINK = 'https://api-kovan.etherscan.io/api?';
export const KOVAN_FROM_BLOCK = '6900000';
export const ACCEPT_PRICE_EVENT = '0x8eb94c6a87f56bd59f4a2a7d571f32a264458ff5b910a34862b9051e5953442d';