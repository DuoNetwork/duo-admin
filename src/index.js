import pf from "./priceSender/priceFeed";
import listenToAcceptPrice from "./listenToAcceptPrice";
import createAccount from "./accounts/createAccounts";
import bitfinexTradeFeedUtil from "./dataFetcher/bitfinex/bitfinexUtil";
import geminiTradeFeedUtil from "./dataFetcher/gemini/geminiUtil";
import krankenTradeFeedUtil from "./dataFetcher/kraken/krakenUtil";
import coinbaseGDAXTradeFeedUtil from "./dataFetcher/gdax/gdaxUtil";

let tool = process.argv[2];

switch (tool) {
	case "pf":
		console.log("starting commitPrice process");
		pf.startFeeding();
		break;
	case "acceptPrice":
		console.log("starting listening to acceptPrice event");
		listenToAcceptPrice.startListening();
		break;
	case "createAccount":
		console.log("starting create accounts");
		createAccount.createAccount(process.argv[3]);
		break;
	case "bitfinex":
		console.log("starting fetchTrade of bitfinex");
		bitfinexTradeFeedUtil.fetchETHTradesByOwnWebSocket();
		break;
	case "gemini":
		console.log("starting fetchTrade of gemini");
		geminiTradeFeedUtil.fetchETHTradesByOwnWebSocket();
		break;
	case "kraken":
		console.log("starting fetchTrade of kraken");
		krankenTradeFeedUtil.startFetching();
		break;

	case "gdax":
		console.log("starting fetchTrade of gdax");
		coinbaseGDAXTradeFeedUtil.startFetching();
		break;
}
