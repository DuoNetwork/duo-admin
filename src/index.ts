import bitfinexUtil from './apis/bitfinexUtil';
import gdaxUtil from './apis/gdaxUtil';
import geminiUtil from './apis/geminiUtil';
import krakenUtil from './apis/krakenUtil';
import ContractUtil from './contractUtil';
import dynamoUtil from './database/dynamoUtil';
import sqlUtil from './database/sqlUtil';
import dbUtil from './dbUtil';
import eventUtil from './eventUtil';
import util from './util';

const tool = process.argv[2];
util.log('tool ' + tool);
const option = util.parseOptions(process.argv);
util.log('using ' + (option.live ? 'live' : 'dev') + ' env and ' + (option.source || 'local node'));
const contractUtil = new ContractUtil(option);
dbUtil.init(option.dynamo);
dynamoUtil.init(
	option.live,
	util.getDynamoRole(tool, option.dynamo),
	util.getStatusProcess(tool, option)
);
if (['bitfinex', 'gemini', 'kraken', 'gdax', 'commit'].includes(tool) && !option.dynamo) {
	const mysqlAuthFile = require('./keys/mysql.json');
	sqlUtil.init(mysqlAuthFile.host, mysqlAuthFile.user, mysqlAuthFile.password);
}

switch (tool) {
	case 'bitfinex':
		util.log('starting fetchTrade of bitfinex');
		bitfinexUtil.fetchTrades();
		break;
	case 'gemini':
		util.log('starting fetchTrade of gemini');
		geminiUtil.fetchTrades();
		break;
	case 'kraken':
		util.log('starting fetchTrade of kraken');
		krakenUtil.startFetching();
		break;
	case 'gdax':
		util.log('starting fetchTrade of gdax');
		gdaxUtil.startFetching();
		break;
	case 'subscribe':
		eventUtil.subscribe(contractUtil, option);
		setInterval(dynamoUtil.insertHeartbeat, 30000);
		break;
	case 'commit':
		util.log('starting commit process');
		contractUtil.commitPrice(option);
		break;
	case 'minutely':
		dynamoUtil
			.readTradeData('GDAX', '2018-06-01-05-54')
			.then(data => console.log(JSON.stringify(data)));
		break;
	default:
		util.log('no such tool ' + tool);
		break;
}
