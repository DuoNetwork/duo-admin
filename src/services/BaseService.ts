import DualClassWrapper from '../../../duo-contract-wrapper/src/DualClassWrapper';
import EsplanadeWrapper from '../../../duo-contract-wrapper/src/EsplanadeWrapper';
import MagiWrapper from '../../../duo-contract-wrapper/src/MagiWrapper';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import { IOption } from '../common/types';
import dbUtil from '../utils/dbUtil';

export default class BaseService {
	public web3Wrapper: Web3Wrapper;
	public tool: string;
	public option: IOption;

	constructor(tool: string, option: IOption) {
		this.option = option;
		this.tool = tool;
		this.web3Wrapper = new Web3Wrapper(null, option.provider, '', option.live);
	}

	public init() {
		return dbUtil.init(this.tool, this.option, this.web3Wrapper);
	}

	public createDuoWrappers(): {[type: string]: {[tenor: string]: DualClassWrapper}} {
		return {
			Beethoven: {
				Perpetual: new DualClassWrapper(
					this.web3Wrapper,
					this.web3Wrapper.contractAddresses.Custodians.Beethoven.Perpetual.custodian.address
				),
				M19: new DualClassWrapper(
					this.web3Wrapper,
					this.web3Wrapper.contractAddresses.Custodians.Beethoven.M19.custodian.address
				)
			},
			Mozart: {
				Perpetual: new DualClassWrapper(
					this.web3Wrapper,
					this.web3Wrapper.contractAddresses.Custodians.Mozart.Perpetual.custodian.address
				),
				M19: new DualClassWrapper(
					this.web3Wrapper,
					this.web3Wrapper.contractAddresses.Custodians.Mozart.M19.custodian.address
				)
			}
		};
	}

	public createMagiWrapper() {
		return new MagiWrapper(
			this.web3Wrapper,
			this.web3Wrapper.contractAddresses.Oracles[0].address
		);
	}

	public createEsplanadeWrapper() {
		return new EsplanadeWrapper(
			this.web3Wrapper,
			this.web3Wrapper.contractAddresses.MultiSigManagers[0].address
		);
	}
}
