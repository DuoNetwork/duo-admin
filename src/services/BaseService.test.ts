import { kovan } from '../../../duo-contract-wrapper/src/contractAddresses';
import BaseService from './BaseService';
jest.mock('../../../duo-contract-wrapper/src/Web3Wrapper', () =>
	jest.fn(() => ({
		contractAddresses: kovan
	}))
);
jest.mock('../../../duo-contract-wrapper/src/DualClassWrapper', () =>
	jest.fn(() => ({
		contract: 'dualClassWrapper'
	}))
);
jest.mock('../../../duo-contract-wrapper/src/EsplanadeWrapper', () =>
	jest.fn(() => ({
		contract: 'EsplanadeWrapper'
	}))
);
jest.mock('../../../duo-contract-wrapper/src/MagiWrapper', () =>
	jest.fn(() => ({
		contract: 'MagiWrapper'
	}))
);

const service = new BaseService('tool', {
	live: false,
	provider: 'provider'
} as any);

test('createDuoWrappers', () => {
	expect(service.createDuoWrappers()).toMatchSnapshot();
});

test('createMagiWrapper', () => {
	expect(service.createMagiWrapper()).toMatchSnapshot();
});

test('createEsplanadeWrapper', () => {
	expect(service.createEsplanadeWrapper()).toMatchSnapshot();
});
