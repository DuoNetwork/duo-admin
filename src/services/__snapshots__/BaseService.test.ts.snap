// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDuoWrappers 1`] = `
Object {
  "Beethoven": Object {
    "M19": Object {
      "contract": "dualClassWrapper",
    },
    "Perpetual": Object {
      "contract": "dualClassWrapper",
    },
  },
  "Mozart": Object {
    "M19": Object {
      "contract": "dualClassWrapper",
    },
    "Perpetual": Object {
      "contract": "dualClassWrapper",
    },
  },
}
`;

exports[`createEsplanadeWrapper 1`] = `
Object {
  "contract": "EsplanadeWrapper",
}
`;

exports[`createMagiWrapper 1`] = `
Object {
  "contract": "MagiWrapper",
}
`;
